llvm/lib/Analysis/DXILResource.cpp

```diff
@@ -823,8 +827,153 @@ DXILBindingMap::findByUse(const Value *Key) const {
 
 //===----------------------------------------------------------------------===//
 
+static bool isUpdateCounterIntrinsic(Function &F) {
+  return F.getIntrinsicID() == Intrinsic::dx_resource_updatecounter;
+}
+
+void DXILResourceCounterDirectionMap::populate(Module &M, DXILBindingMap &DBM) {
+  std::vector<std::tuple<dxil::ResourceBindingInfo, ResourceCounterDirection,
+                         const Function *, const CallInst *>>
+      DiagCounterDirs;
+
+  for (Function &F : M.functions()) {
+    if (!isUpdateCounterIntrinsic(F))
+      continue;
+
+    for (const User *U : F.users()) {
+      const CallInst *CI = dyn_cast<CallInst>(U);
+      assert(CI && "Users of dx_resource_updateCounter must be call instrs");
+
+      // Determine if the use is an increment or decrement
+      Value *CountArg = CI->getArgOperand(1);
+      ConstantInt *CountValue = dyn_cast<ConstantInt>(CountArg);
+      int64_t CountLiteral = CountValue->getSExtValue();
+
+      ResourceCounterDirection Direction = ResourceCounterDirection::Unknown;
+      if (CountLiteral > 0)
+        Direction = ResourceCounterDirection::Increment;
+      if (CountLiteral < 0)
```

Comment:
Doesn't really matter but this might as well be `else if`

Reply:
imo `else` (and extra so for `else if`)  harms readability because you have to keep  the context of the negation  of the previous `if` in your head to tease out the logic. Happy to make the change but my default programming habit is to minimize the use of else as much as possible without writing extra wonky code

Reply:
latest change ends up fully deleting one of the `if`s

Reply:
I see you have already resolved this differently. :) I just wanted to add that for me the double `if` makes it actually "harder" to read because one needs to read & evaluate the condition twice, and then wonder "why is this not an `else`"?

Another option here might be 
`Direction = CountLiteral > 0 ? ResourceCounterDirection::Increment : ResourceCounterDirection::Decrement;`

(No changes needed, I was just writing this comment when you pushed the update.)